src/mds/MDSDaemon.cc

```diff
@@ -896,6 +896,8 @@ void MDSDaemon::suicide()
 
   clean_up_admin_socket();
 
+  mds_lock.unlock();
```

Comment:
(I was hoping the tracker ticket would also explain this:) so can you elaborate where exactly the deadlock is? We often call these `beacon.` methods with the `mds_lock` held.

Reply:
I'll amend the commit message with an explanation.

Reply:
Commit message (and PR text) amended.

Reply:
Thanks for explaining, that makes sense.

I think the danger here is that these methods intend to sit on the `mds_lock` precisely to prevent any work from being done in the background while waiting for the beacon reply. However, I don't _think_ that's a problem since `MDSDaemon::ms_dispatch2` drops out if the state has been changed to `_DNE`:

https://github.com/ceph/ceph/blob/07555ce91d0d80332e86ff0d5faff73500af830d/src/mds/MDSDaemon.cc#L1000-L1004

So I think this is safe but only after changing the want state and just before `send_and_wait`.

Note: `MDSRank::damaged` needs adjusted too.